src/_posts/2019-10-30-crafting-websites-using-finite-state-machines-part-I.md

```diff
@@ -0,0 +1,403 @@
+---
+author: Dan Bunea
+layout: post
+asset-type: post
+title: "Crafting web apps using finite state machines - Part 1"
+date: 2019-10-30 00:00:00
+description: A practical approach to crafting web applications, with the goals of breaking the complexity down (state machines, separation of concerns) and building robustness right into the system (atomicity with rollback).
+image:
+    src: /assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/controller%20to%20diagram.png
+    attribution:
+
+abstract: Crafting web apps using finite state machines, separation of concerns and atomicity.
+tags:
+- functional programming
+- ClojureScript
+- finite state machines
+- model view controller
+---
+
+# Crafting web apps using finite state machines - Part 1
+
+
+[Code on Gitlab](https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps)
+
+**Table of contents**
+
+- PROBLEM: How should we start crafting a web app?
+- SOLUTION
+  * THEORY
+  * Strategy
+    + Goals
+  * Tactics
+    + Patterns
+    + Techniques
+  * PRACTICE
+    + Step 1 Draw the diagram of the UI state machine
+    + Step 2: Domain modelling
+    + Step 3: Solve the data problem
+    + Step 4: Build the UI
+  * Goals revised
+  * Part II - what will be discussed
+
+
+# PROBLEM: How should we start crafting a web app?
+
+The requirements come, we have some UI mockups and now we need to start developing a web app. What is the first step we need to do?
+
+Let's go back a bit. If this was not a specific web app but rather a generic one, then:
+
+   1. What would our strategy be like?
+   2. What would be our goals?
+   3. Which steps would we have to do?
+   4. Which patterns could help us?
+   5. What tactics could we use?
+
+In the following lines I will try to address exactly this questions. We'll start with a little bit of theory, then we'll put it in practice:
+
+<a name="solution"/>
+
+# SOLUTION
+
+## THEORY
+
+## Strategy
+
+
+We'll have to think a bit strategically and a little bit tactically. For the strategy, what would be our goals:
+
+### Goals
+
+* Break the complexity down (state machines, separation of concerns)
+* Build robustness into the system (atomicity with rollback)
+
+Ok, but how can this be done?
+
+Well, we'll do it in 4 Steps
+1. Finite state machine diagram (state/transitions starting from the UI)
+2. States (domain modelling for the states)
+3. Transitions (TDD the controller)
+4. Presentation/UI for the states (TDD the view)
+
+Steps don't need to be sequential, in fact it is recommended to do the last two in parallel.
+
+## Tactics
+
+### Patterns 
```

Comment:
Are you going to expand this and the next section? Because at the moment I would just remove the ### lines

Reply:
Yes, this is one of the part I will expand in part II